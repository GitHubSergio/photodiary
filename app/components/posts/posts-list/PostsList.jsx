// Imports
import React, { useEffect } from 'react';
import { FlatList } from 'react-native';
import { useSelector, useDispatch } from 'react-redux';
import { useIsFocused, useNavigation } from '@react-navigation/native';
import postsListStyles from './postsListStyles';

// Actions
import { getAllPosts } from '../../../redux/middlewares/postsMiddleWare';
import {
  clearPostsSuccess,
  clearDeletePostSuccess,
  clearUpdatePostSuccess,
} from '../../../redux/actions/postsActions';

// Components
import PostsListItem from '../posts-list-item/PostsListItem';
import Loader from '../../shared/loader/Loader';
import NoPosts from '../no-posts/NoPosts';
import Button from '../../shared/buttons/Button';

const ITEM_HEIGHT = 200;
const extractKey = ({ postId }) => postId;

const PostsList = () => {
  const { email } = useSelector((state) => state.user.userDetails);
  const { allPosts } = useSelector((state) => state.posts);
  const { isFetching } = useSelector((state) => state.posts);
  const { createPostSuccess } = useSelector((state) => state.posts);
  const { deletePostSuccess } = useSelector((state) => state.posts);
  const { updatePostSuccess } = useSelector((state) => state.posts);

  const dispatch = useDispatch();
  const navigation = useNavigation();
  const isFocused = useIsFocused();

  useEffect(() => {
    if (email) {
      dispatch(getAllPosts());
      console.log('USEEFFECT ONE');
      // TODO implement https://reactnavigation.org/docs/use-scroll-to-top
    }
  }, [dispatch, email]);

  useEffect(() => {
    if (isFocused && createPostSuccess) {
      console.log('USEEFFECT IF ONE');
      dispatch(clearPostsSuccess(0));
      dispatch(getAllPosts());
    }

    if (updatePostSuccess) {
      console.log('USEEFFECT IF TWO');
      dispatch(clearUpdatePostSuccess(0));
      dispatch(getAllPosts());
    }

    if (isFocused && deletePostSuccess) {
      console.log('USEEFFECT IF THREE');
      dispatch(clearDeletePostSuccess(0));
      dispatch(getAllPosts());
    }
  }, [
    isFocused,
    createPostSuccess,
    deletePostSuccess,
    updatePostSuccess,
    dispatch,
  ]);

  return isFetching ? (
    <Loader />
  ) : (
    <>
      {allPosts.length > 0 ? (
        <FlatList
          testID={'listPosts'}
          data={allPosts}
          removeClippedSubviews
          renderItem={({ item }) => (
            <PostsListItem
              title={item.title}
              description={item.description}
              postId={item.postId}
              imageUri={item.imageUri}
            />
          )}
          keyExtractor={extractKey}
          getItemLayout={(data, index) => ({
            length: ITEM_HEIGHT,
            offset: ITEM_HEIGHT * index,
            index,
          })}
        />
      ) : (
        <NoPosts />
      )}
      <Button
        label="+"
        onPress={() => navigation.navigate('AddNewPostScreen')}
        style={postsListStyles.addButton}
        textStyle={postsListStyles.addButtonText}
      />
    </>
  );
};

export default PostsList;
