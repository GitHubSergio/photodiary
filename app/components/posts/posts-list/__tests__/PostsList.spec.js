/* eslint-disable import/no-extraneous-dependencies */
import React from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { cleanup, render } from '@testing-library/react-native';
import reduxMockedStore from '../../../../utils/redux-state';
import PostsList from '../PostsList';

jest.mock('@react-navigation/native', () => ({
  useNavigation: () => ({ goBack: jest.fn() }),
  useIsFocused: () => ({ isFocused: true }),
}));

jest.mock('react-redux', () => {
  return {
    useSelector: jest.fn(),
    useDispatch: jest.fn(),
  };
});

jest.mock('@react-native-firebase/auth', () => ({}));

jest.mock('@react-native-firebase/firestore', () => ({}));

jest.mock('@react-native-firebase/storage', () => ({}));

describe('<PostsList/>', () => {
  describe('@render', () => {
    let wrapper;

    afterEach(() => {
      jest.resetAllMocks();
      cleanup();
    });

    it('should render the Loader component when loading the list', async () => {
      useSelector.mockImplementation(() => ({
        accessToken: '12345',
        isFetching: true,
      }));
      const getAllPosts = useDispatch.mockImplementationOnce(() => jest.fn());
      wrapper = render(<PostsList />);
      expect(getAllPosts).toHaveBeenCalledTimes(1);
      expect(wrapper.container).toMatchSnapshot();
    });

    it('should render the list of posts', async () => {
      useSelector.mockImplementation(() => ({
        accessToken: '12345',
        allPosts: reduxMockedStore.posts.allPosts,
      }));
      useDispatch.mockImplementationOnce(() => jest.fn());
      wrapper = render(<PostsList />);
      expect(wrapper.container).toMatchSnapshot();
    });

    it('should render the NoPosts component if not posts are found', async () => {
      useSelector.mockImplementation(() => ({
        accessToken: '123',
        allPosts: [],
      }));
      useDispatch.mockImplementationOnce(() => jest.fn());
      wrapper = render(<PostsList />);
      expect(wrapper.container).toMatchSnapshot();
    });

    it('should dispatch the clearPostSuccess', () => {
      useSelector.mockImplementation(() => ({
        accessToken: '123',
        allPosts: reduxMockedStore.posts.allPosts,
        success: true,
      }));
      const clearPostsSuccess = useDispatch.mockImplementationOnce(() =>
        jest.fn(),
      );
      wrapper = render(<PostsList />);
      expect(clearPostsSuccess).toHaveBeenCalledTimes(1);
    });
  });
});
