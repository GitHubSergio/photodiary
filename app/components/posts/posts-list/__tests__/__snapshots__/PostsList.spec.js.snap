// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<PostsList/> @render should render the Loader component when loading the list 1`] = `
<View
  collapsable={true}
  pointerEvents="box-none"
  style={
    Object {
      "flex": 1,
    }
  }
>
  <View
    testID="loader"
  >
    <ActivityIndicator
      animating={true}
      color="#0000ff"
      hidesWhenStopped={true}
      size="large"
    />
  </View>
</View>
`;

exports[`<PostsList/> @render should render the NoPosts component if not posts are found 1`] = `
<View
  collapsable={true}
  pointerEvents="box-none"
  style={
    Object {
      "flex": 1,
    }
  }
>
  <View
    style={
      Object {
        "flex": 1,
        "justifyContent": "center",
      }
    }
  >
    <Text
      style={
        Object {
          "color": "#688598",
          "fontSize": 50,
          "textAlign": "center",
        }
      }
    >
      NO POSTS YET
    </Text>
  </View>
  <
    style={
      Array [
        Object {
          "alignItems": "center",
          "backgroundColor": "#2D2A4A",
          "borderRadius": 10,
          "height": 40,
          "justifyContent": "center",
        },
        Object {
          "borderRadius": 100,
          "bottom": 30,
          "height": 50,
          "position": "absolute",
          "right": 10,
          "shadowColor": "#2D2A4A",
          "shadowOffset": Object {
            "height": 2,
            "width": 0,
          },
          "shadowOpacity": 0.5,
          "shadowRadius": 3.84,
          "width": 50,
        },
      ]
    }
  >
    <Text
      accessibilityLabel="submit-button"
      style={
        Array [
          Object {
            "color": "#F4F5F8",
            "fontSize": 20,
            "fontWeight": "bold",
          },
          Object {
            "fontSize": 30,
          },
        ]
      }
    >
      +
    </Text>
  </>
</View>
`;

exports[`<PostsList/> @render should render the list of posts 1`] = `
<View
  collapsable={true}
  pointerEvents="box-none"
  style={
    Object {
      "flex": 1,
    }
  }
>
  <ScrollView
    data={
      Array [
        Object {
          "comments": 0,
          "createdAt": "2020-02-20T23:08:31.704Z",
          "description": "Bangkok, Thailand’s capital, is a large city known for ornate shrines and vibrant street life. The boat-filled Chao Phraya River feeds its network of canals, flowing past the Rattanakosin royal district, home to opulent Grand Palace and its sacred Wat Phra Kaew Temple. Nearby is Wat Pho Temple with an enormous reclining Buddha and, on the opposite shore, Wat Arun Temple with its steep steps and Khmer-style spire.",
          "docRef": "0264d0b6-a99f-421f-8d02-f87008888de8",
          "image": "imgUrl",
          "likes": 0,
          "title": "Bangkok",
          "userId": "YVHTtPKcyzReNwHNv3oyV9ArK8n2",
        },
        Object {
          "comments": 0,
          "createdAt": "2020-02-20T23:20:47.603Z",
          "description": "Phuket City, on Phuket Island, is the capital of Thailand’s Phuket Province. In the Old Town, Thalang Road is lined with colorful 19th-century shophouses and Sino-Portuguese buildings. Built in 1903 by a wealthy tin merchant, Baan Chinpracha mansion has Italian floor tiles, shuttered windows and antique furniture. Set in a 1930s manor, the Thai Hua Museum has exhibits on Phuket’s culture and history.",
          "docRef": "52e6b747-ec3d-4bb3-b445-49cabbc85bca",
          "image": "imgUrl",
          "likes": 0,
          "title": "Phuket ",
          "userId": "YVHTtPKcyzReNwHNv3oyV9ArK8n2",
        },
        Object {
          "comments": 0,
          "createdAt": "2020-02-20T23:07:56.308Z",
          "description": "Ko Samui, Thailand’s second largest island, lies in the Gulf of Thailand off the east coast of the Kra Isthmus. It's known for its palm-fringed beaches, coconut groves and dense, mountainous rainforest, plus luxury resorts and posh spas. The landmark 12m-tall golden Big Buddha statue at Wat Phra Yai Temple is located on a tiny island connected to Ko Samui by a causeway.",
          "docRef": "93caf8c8-0bf5-4625-9ffa-bc8b8ed31a56",
          "image": "imgUrl",
          "likes": 0,
          "title": "Koh Samui",
          "userId": "YVHTtPKcyzReNwHNv3oyV9ArK8n2",
        },
      ]
    }
    disableVirtualization={false}
    horizontal={false}
    initialNumToRender={10}
    maxToRenderPerBatch={10}
    onEndReachedThreshold={2}
    removeClippedSubviews={true}
    scrollEventThrottle={50}
    stickyHeaderIndices={Array []}
    testID="listPosts"
    updateCellsBatchingPeriod={50}
    viewabilityConfigCallbackPairs={Array []}
    windowSize={21}
  >
    <View
      style={null}
    >
      <
        style={
          Object {
            "alignItems": "center",
            "backgroundColor": "#F4F5F8",
            "flex": 1,
            "flexDirection": "row",
            "height": 150,
            "marginBottom": 10,
            "shadowColor": "#2D2A4A",
            "shadowOffset": Object {
              "height": 2,
              "width": 0,
            },
            "shadowOpacity": 0.5,
            "shadowRadius": 3.84,
          }
        }
      >
        <View
          style={
            Object {
              "alignItems": "center",
              "flex": 1,
              "justifyContent": "center",
            }
          }
        >
          <Image
            resizeMode="cover"
            source={
              Object {
                "uri": undefined,
              }
            }
            style={
              Object {
                "height": 150,
                "width": 150,
              }
            }
          />
        </View>
        <View
          style={
            Object {
              "flex": 1,
            }
          }
        >
          <View
            style={
              Object {
                "alignItems": "center",
              }
            }
          >
            <Text
              ellipsizeMode="tail"
              numberOfLines={1}
              style={
                Object {
                  "fontSize": 16,
                  "fontWeight": "bold",
                }
              }
            >
              Bangkok
            </Text>
          </View>
          <View
            style={
              Object {
                "flex": 1,
                "paddingLeft": 0,
              }
            }
          >
            <Text
              ellipsizeMode="tail"
              numberOfLines={5}
            >
              Bangkok, Thailand’s capital, is a large city known for ornate shrines and vibrant street life. The boat-filled Chao Phraya River feeds its network of canals, flowing past the Rattanakosin royal district, home to opulent Grand Palace and its sacred Wat Phra Kaew Temple. Nearby is Wat Pho Temple with an enormous reclining Buddha and, on the opposite shore, Wat Arun Temple with its steep steps and Khmer-style spire.
            </Text>
          </View>
        </View>
      </>
    </View>
    <View
      style={null}
    >
      <
        style={
          Object {
            "alignItems": "center",
            "backgroundColor": "#F4F5F8",
            "flex": 1,
            "flexDirection": "row",
            "height": 150,
            "marginBottom": 10,
            "shadowColor": "#2D2A4A",
            "shadowOffset": Object {
              "height": 2,
              "width": 0,
            },
            "shadowOpacity": 0.5,
            "shadowRadius": 3.84,
          }
        }
      >
        <View
          style={
            Object {
              "alignItems": "center",
              "flex": 1,
              "justifyContent": "center",
            }
          }
        >
          <Image
            resizeMode="cover"
            source={
              Object {
                "uri": undefined,
              }
            }
            style={
              Object {
                "height": 150,
                "width": 150,
              }
            }
          />
        </View>
        <View
          style={
            Object {
              "flex": 1,
            }
          }
        >
          <View
            style={
              Object {
                "alignItems": "center",
              }
            }
          >
            <Text
              ellipsizeMode="tail"
              numberOfLines={1}
              style={
                Object {
                  "fontSize": 16,
                  "fontWeight": "bold",
                }
              }
            >
              Phuket 
            </Text>
          </View>
          <View
            style={
              Object {
                "flex": 1,
                "paddingLeft": 0,
              }
            }
          >
            <Text
              ellipsizeMode="tail"
              numberOfLines={5}
            >
              Phuket City, on Phuket Island, is the capital of Thailand’s Phuket Province. In the Old Town, Thalang Road is lined with colorful 19th-century shophouses and Sino-Portuguese buildings. Built in 1903 by a wealthy tin merchant, Baan Chinpracha mansion has Italian floor tiles, shuttered windows and antique furniture. Set in a 1930s manor, the Thai Hua Museum has exhibits on Phuket’s culture and history.
            </Text>
          </View>
        </View>
      </>
    </View>
    <View
      style={null}
    >
      <
        style={
          Object {
            "alignItems": "center",
            "backgroundColor": "#F4F5F8",
            "flex": 1,
            "flexDirection": "row",
            "height": 150,
            "marginBottom": 10,
            "shadowColor": "#2D2A4A",
            "shadowOffset": Object {
              "height": 2,
              "width": 0,
            },
            "shadowOpacity": 0.5,
            "shadowRadius": 3.84,
          }
        }
      >
        <View
          style={
            Object {
              "alignItems": "center",
              "flex": 1,
              "justifyContent": "center",
            }
          }
        >
          <Image
            resizeMode="cover"
            source={
              Object {
                "uri": undefined,
              }
            }
            style={
              Object {
                "height": 150,
                "width": 150,
              }
            }
          />
        </View>
        <View
          style={
            Object {
              "flex": 1,
            }
          }
        >
          <View
            style={
              Object {
                "alignItems": "center",
              }
            }
          >
            <Text
              ellipsizeMode="tail"
              numberOfLines={1}
              style={
                Object {
                  "fontSize": 16,
                  "fontWeight": "bold",
                }
              }
            >
              Koh Samui
            </Text>
          </View>
          <View
            style={
              Object {
                "flex": 1,
                "paddingLeft": 0,
              }
            }
          >
            <Text
              ellipsizeMode="tail"
              numberOfLines={5}
            >
              Ko Samui, Thailand’s second largest island, lies in the Gulf of Thailand off the east coast of the Kra Isthmus. It's known for its palm-fringed beaches, coconut groves and dense, mountainous rainforest, plus luxury resorts and posh spas. The landmark 12m-tall golden Big Buddha statue at Wat Phra Yai Temple is located on a tiny island connected to Ko Samui by a causeway.
            </Text>
          </View>
        </View>
      </>
    </View>
  </ScrollView>
  <
    style={
      Array [
        Object {
          "alignItems": "center",
          "backgroundColor": "#2D2A4A",
          "borderRadius": 10,
          "height": 40,
          "justifyContent": "center",
        },
        Object {
          "borderRadius": 100,
          "bottom": 30,
          "height": 50,
          "position": "absolute",
          "right": 10,
          "shadowColor": "#2D2A4A",
          "shadowOffset": Object {
            "height": 2,
            "width": 0,
          },
          "shadowOpacity": 0.5,
          "shadowRadius": 3.84,
          "width": 50,
        },
      ]
    }
  >
    <Text
      accessibilityLabel="submit-button"
      style={
        Array [
          Object {
            "color": "#F4F5F8",
            "fontSize": 20,
            "fontWeight": "bold",
          },
          Object {
            "fontSize": 30,
          },
        ]
      }
    >
      +
    </Text>
  </>
</View>
`;
